Source/NSStoryboard.m

```diff
@@ -0,0 +1,975 @@
+
+- (instancetype) init
+{
+  self = [super init];
+  if (self != nil)
+    {
+      _storyboardName = nil;
+    }
+  return self;
+}
+
+- (NSString *) storyboardName
+{
+  return _storyboardName;
+}
+
+- (void) setStoryboardName: (NSString *)name
+{
+  ASSIGNCOPY(_storyboardName, name);
+}
+
+- (id) copyWithZone: (NSZone *)z
+{
+  NSControllerPlaceholder *c = [[NSControllerPlaceholder allocWithZone: z] init];
+  [c setStoryboardName: _storyboardName];
+  return c;
+}
+
+- (instancetype) initWithCoder: (NSCoder *)coder
+{
+  self = [super init];
+  if ([coder allowsKeyedCoding])
+    {
+      if ([coder containsValueForKey: @"NSStoryboardName"])
+        {
+          [self setStoryboardName: [coder decodeObjectForKey: @"NSStoryboardName"]];
+        }
+    }
+  return self;
+}
+
+- (void) encodeWithCoder: (NSCoder *)coder
+{
+  // this is never encoded directly...
+}
+
+- (id) instantiate
+{
+  _storyboard = [NSStoryboard storyboardWithName: _storyboardName
+                                          bundle: [NSBundle mainBundle]];
+  return [_storyboard instantiateInitialController];
+}
+
+@end
+
+@implementation NSStoryboard
+
+- (NSXMLElement *) _createCustomObjectWithId: (NSString *)ident
+                                   userLabel: (NSString *)userLabel
+                                 customClass: (NSString *)className
+{
+  NSXMLElement *customObject =
+    [[NSXMLElement alloc] initWithName: @"customObject"];
+  NSXMLNode *idValue =
+    [NSXMLNode attributeWithName: @"id"
+                     stringValue: ident];
+  NSXMLNode *usrLabel =
+    [NSXMLNode attributeWithName: @"userLabel"
+                     stringValue: userLabel];
+  NSXMLNode *customCls =
+    [NSXMLNode attributeWithName: @"customClass"
+                     stringValue: className];
+  
+  [customObject addAttribute: idValue];
+  [customObject addAttribute: usrLabel];
+  [customObject addAttribute: customCls];
+
+  AUTORELEASE(customObject);
+  
+  return customObject;
+}
+
+- (void) _processStoryboard: (NSXMLDocument *)storyboardXml
+{
+  NSArray *docNodes = [storyboardXml nodesForXPath: @"document" error: NULL];
+
+  if ([docNodes count] > 0)
+    {
+      NSXMLElement *docNode = [docNodes objectAtIndex: 0];
+      NSArray *array = [docNode nodesForXPath: @"//scene" error: NULL];
+      NSEnumerator *en = [array objectEnumerator];
+      NSXMLElement *e = nil;  
+      NSString *customClassString = nil;
+      
+      // Set initial view controller...
+      ASSIGN(_initialViewControllerId, [[docNode attributeForName: @"initialViewController"] stringValue]);             
+      _scenesMap = [[NSMutableDictionary alloc] initWithCapacity: [array count]];
+      _controllerMap = [[NSMutableDictionary alloc] initWithCapacity: [array count]];
+      _documentsMap = [[NSMutableDictionary alloc] initWithCapacity: [array count]];
+      
+      while ((e = [en nextObject]) != nil)
+        {
+          NSXMLElement *doc = [[NSXMLElement alloc] initWithName: @"document"];
```

Comment:
This method is way too long. Could you at least move the code in this loop into a helper method?

Reply:
This is part of my planned refactoring if we can defer this. I would appreciate it.  I am planning on moving all of the transformation logic out of this file.

Reply:
Deferred until refactor.

Reply:
IN software engineering we call this the "big refactor in the sky" or "grand rewrite in the sky". Feel free to check this out.

Reply:
LOL, true.  It's been done.  I am a Computer Scientist, so it's cool. ;)